### ISSUE TITLE
FPU state issues

### ISSUE THREAD

**Issue by bartoldeman** [dosemu2/dosemu2#572]
In the end I did not touch the FPU in #571 but I did see a few issues (also when trying to revive full cpuemu, that is still a local patchset).

what we have for V86
FPU state is restored/saved in vm86_fpu_state, however:
a. for KVM this is not necessary because it's got its own separate FPU state
b. it is not necessary for vm86() if dosemu itself does not use the FPU between vm86 calls (where does it do so nowadays?)
c. using the simulator the state is simulated, no need to save and restore
d. using the jit, the jit knows which code sections access the FPU so can skip save/restore for non-FPU using code.

for DPMI:
a. KVM and cpuemu are as above
b. native saves/restores FPU state naturally via signals

for vm86->DPMI:
a. state is copied, but irrelevant if both use KVM (where state is then kept inside the VM)

for DPMI->vm86
a. I believe (if not using KVM for both) the state is lost and it will pick up FPU state from the last vm86 FPU save unless I missed something

for i386 only:
struct _libc_fpstate is shorter than struct _fpstate... which means SSE state is not saved/restored if that is used, and only the classic x87 registers are done.

I don't think the DPMI spec has much to say about this subject... however saving/restoring FPU state is not cheap so if it's not necessary it should not be done.

**Comment by stsp** [comment]
> b. it is not necessary for vm86() if dosemu itself does not use the
> FPU between vm86 calls (where does it do so nowadays?)

Just where it was initially - in the sound code.
You can grep it for `double`.
But the problem is not only that: any lib
that we link in, may use it just behind our
back. I think the ideas to not restore it,
come from the stone age where dosemu
code was pretty much standalone, w/o
many libs.

> struct _libc_fpstate is shorter than struct _fpstate...
> which means SSE state is not saved/restored if that
> is used, and only the classic x87 registers are done.

If it is so, this is a problem.
I remember djgpp-compiled FP math uses SSE.

> don't think the DPMI spec has much to say about this subject...
> however saving/restoring FPU state is not cheap so if it's not necessary it should not be done.

In what case?
When switching from anything to dosemu, it is
enough to reset the FPU state. We don't restore
it IIRC. When switching to DPMI, in case of the
KVM, as you said, its not needed at all.
So what is left? Native DPMI? I think we'll just
undefault it for security reasons.

**Comment by stsp** [comment]
It seems I am no longer receiving the
github notifications by e-mail. :(
Is it only me?

**Comment by bartoldeman** [comment]
Not sure if you are using your sourceforge forwarding address to get notifications, but SF has been plagued by DDOS attacks. Github too, but github survived.

I'll send patches addressing all the FPU issues. Indeed native DPMI and good old vm86() need save/restore for sure, and cpuemu conditionally since the JIT uses the real FPU stack.

**Comment by stsp** [comment]
Good point!
Changed the address, lets see.
Thanks so far. :)

**Comment by bartoldeman** [comment]
After looking at the DPMI code a bit further I think a choice needs to be made. Either:

1. FPU state is never copied, and left on the signal stack. dpmi_init() puts NULL in scp->fpregs so that sigreturn() initializes the FPU, and similar for when a nested DPMI client just ended. This is sufficient to avoid dosemu corrupting dpmi client FPU state (ever since the removal of direct DPMI switching). copy_context would be somewhat faster without FPU state copying. I don't think there is any expectation of DPMI clients inheriting vm86 FPU state since DOS extenders and DOS C runtimes typically initialize the FPU anyway (see e.g. http://dos32a.narechk.net/manual/html/tech/2.htm)

2. FPU context is always copied when going from DPMI to vm86 and back too, not just at dpmi_init(). This is the most comprehensive, but I'd be surprised if any DPMI client depends on this.

3. Leave as is. dpmi_init copies vm86 FPU context, but is otherwise per client, and FPU context is never copied back to vm86.

I vote for 1 as it's simplest and sufficient to my knowledge, but could be convinced for 2. 3 doesn't make sense to me.

**Comment by stsp** [comment]
> FPU state is never copied, and left on the signal stack. dpmi_init() puts NULL
> in scp->fpregs so that sigreturn() initializes the FPU

I don't think I have time right now to look
why it is so. My assumption was that if you
pass NULL, then it will just not change the
FPU state. So if it does otherwise, I believe
this behaviour is inherently unportable. To
what exactly values does it initialize an FPU
state when you pass NULL?

> 3 doesn't make sense to me.

But could you please explain why? :)
The problem with 2 that I can think of, is
when the DPMI client terminates. It may
also terminate by an FPU exception. It
would really be strange to propagate this
back to vm86. I.e. if you copy the state
back to vm86, it will inevitably be inherited
by the unrelated realmode progs that run
after DPMI prog exited.
The fact that 3 copies FPU context from
vm86 to DPMI is not important. I was just
seeking for the good initialization source.
If we modify 3 to use some clean FPU
context (cooked by FPU reset procedure),
then will it look better in your opinion? If
so, this should be case 4. :)

**Comment by bartoldeman** [comment]
> My assumption was that if you
pass NULL, then it will just not change the
FPU state. So if it does otherwise, I believe
this behaviour is inherently unportable. To
what exactly values does it initialize an FPU
state when you pass NULL?

/usr/include/asm/sigcontext.h says: struct _fpstate         *fpstate;       /* Zero when no FPU context */

In older kernels with lazy fpu context switching the kernel disabled the FPU essentially, setting the TS flag in CR0 if not already set. Having NULL in FPU state was normal, if a signal was delivered to a non-FPU using process.

Newer kernels do eager fpu context switching, as most processes use SSE anyway (via memcpy etc), and lazy was quite slow. In that case if you pass NULL it invokes fpu__clear() in the kernel, doing effectively an FPU init (by doing frstor or similar on a structure with cooked default values, from fpstate_init_fstate() etc)

None of this is portable. The gregset_t structure from mcontext_t is somewhat portable (though POSIX specifies mcontext_t as opaque, it is specified in the sysv i386 abi document, http://www.sco.com/developers/devspecs/abi386-4.pdf pages 206 and 207), but fpregset_t on Linux is completely different, a pointer instead of a struct.

Ideally I'd like to decouple struct _libc_fpstate/_fpstate/fpregset_t from loadfpstate/savefpstate, with all the SSE/AVX/etc variations out there, and _libc_fpstate missing the SSE bits for i386. One can do that by initializing a clean FPU state (as in case 4)  in  indirect_dpmi_transfer() before the pthread_kill(). Then dpmi_switch_sa does not need to touch FPU state on the signal stack at all. Then after the DPMI client exits, dosemu can do 
fesetenv(&dosemu_fenv) in indirect_dpmi_transfer() after the pthread_kill(). 

I also think a fesetenv(&dosemu_fenv) is missing during DPMI execution; this just means that feenableexcept(FE_DIVBYZERO | FE_OVERFLOW) is not active for that part of the code yet if I understand things correctly, as it is only set after the execution of some vm86 code, not when entering a signal handler. Just need to figure out where the best place would be to put it.

**Comment by stsp** [comment]
Does the kernel do fpu_clear also when entering
the signal handler, or should we do that by hands?

> Then dpmi_switch_sa does not need to touch FPU state on the signal stack at all.

Is this because we do no longer return to dosemu
via sigreturn, except when DPMI client exits?

> Then after the DPMI client exits, dosemu can do
> fesetenv(&dosemu_fenv) in indirect_dpmi_transfer() after the pthread_kill().

Likely, but what problem does this solve?
Is it any better than to just replace things on
a signal stack when client exits?

**Comment by bartoldeman** [comment]
> Does the kernel do fpu_clear also when entering
the signal handler, or should we do that by hands?

I am not sure. For very old hardware it's using FNSAVE which clears the FPU, but newer hardware I should check with a test program. I suspect not and we may need to do it by hand.

> Is this because we do no longer return to dosemu
via sigreturn, except when DPMI client exits?

That is really what allows us to keep the FPU state on the signal stack and not copy it to and from what is allocated by DPMI_CLIENT.fpu_state.fpregs.

> Likely, but what problem does this solve?
Is it any better than to just replace things on
a signal stack when client exits?

Basically I'd like to treat the FPU state on the signal stack as an opaque structure of unknown size, which is more portable (I mean between MUSL, glibc mcontext_t, asm/sigcontext.h, for i386 and "struct _xstate"). I'll only make an exception for print_exception_info() which is read/only anyway.

**Comment by stsp** [comment]
> Basically I'd like to treat the FPU state on the signal stack as
> an opaque structure of unknown size, which is more portable

Maybe this makes some sense so far.
But this is a limitation too, so once the dust
around `struct _fpstate` settles, I'd rather
get back to the current technique.
For example you won't be able to switch
between DPMI clients consistently. IIRC
there were some cases when the interrupt
should be delivered not to the current
client (but I forgot the details).
So if you do that, I won't stop you, but
please, at least make this revertable somehow,
maybe with ifdefs, or just don't change the
code too much, to leave the revert opportunity. :)

**Comment by stsp** [comment]
Bart, do we still need this ticket?

**Comment by bartoldeman** [comment]
yes. haven't had much time to work on it, but the i386 sse state fix is still pending.

**Comment by stsp** [comment]
Hi Bart, as you are working on an
FPU right now, can you also revisit
this ticket? You said there are some
FPU problems here... Which ones?

**Comment by stsp** [comment]
I think most of FPU problems were
solved. Let me know if not.
